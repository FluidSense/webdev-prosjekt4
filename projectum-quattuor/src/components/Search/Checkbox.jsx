import React,{Component}  from 'react'

class Checkbox extends Component {
    constructor() {
        super();
        this.state = {
            isChecked: false,
        }
    }

render(){
    const { label } = this.props;
    const { isChecked } = this.state;
    return(
        <div className="checkbox">
            <label>
                <input
                    type="checkbox"
                    value={label} 
                    checked={isChecked} 
                    onChange
                
                    />
                {label}
            </label>
        </div>
    )
}

}

export default Checkbox;