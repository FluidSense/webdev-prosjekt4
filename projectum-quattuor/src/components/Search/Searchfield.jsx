import React from 'react';

class Searchfield extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      searchValue: '',
    };
  }

  handleChange(event) {
    this.setState({ searchValue: event.target.value });
  }

  render() {
    return (
      <div className="search_wrapper">
        <label htmlFor="searchBar">
          <input type="text" id="searchbar" className="search_bar" value={this.state.searchValue} onChange={e => this.handleChange(e)} />
        </label>
        <div>
          <input type="submit" className="search_button" value="May the Force be with you." onClick={() => this.props.updateSearchValue(this.state.searchValue)} />
        </div>
      </div>
    );
  }
}

export default Searchfield;
