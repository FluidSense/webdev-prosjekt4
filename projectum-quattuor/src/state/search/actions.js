/* Data fetching for content helper functions */
export const updateSearchContentSet = param => ({ type: 'FETCHED_RESULTS_RECEIVED', payload: param });
export const updateSearchContentErrorState = param => ({ type: 'FETCH_CONTENT_ERROR_OCCURRED', payload: param });
export const updateSearchContentLoadingState = (loading, loadingUrl) => ({ type: 'FETCH_CONTENT_LOADING_IN_PROGRESS', payload: { loading, loadingUrl } });

/* Data fetching for search history updater */
export const updateSearchHistorySet = param => ({ type: 'FETCHED_HISTORY_RECEIVED', payload: param });

// Fetches data from API
export const fetchSearchData = fetchUrl => (dispatch) => {
  // Updates loading state and informs reducer which url is being fetched
  dispatch(updateSearchContentLoadingState(true, fetchUrl));

  return fetch(fetchUrl)
    .then((response) => {
      if (response.ok) {
        return response.json();
      }
      return false;
    }, (error) => { // Pushes error
      dispatch(updateSearchContentErrorState(error.message));
      dispatch(updateSearchContentLoadingState(false, fetchUrl));
    })
    .then((data) => {
      if (data !== undefined) { // Prevents JSON data to be updated if data does not exist
        dispatch(updateSearchContentSet(data));
      }
    });
};

export const fetchSearchHistory = fetchUrl => (dispatch) => {
  return fetch(fetchUrl)
    .then((response) => {
      if (response.ok) {
        return response.json();
      }
      return false;
    })
    .then((data) => {
      if (data !== undefined) { // Prevents JSON data to be updated if data does not exist
        dispatch(updateSearchHistorySet(data));
      }
    });
};
