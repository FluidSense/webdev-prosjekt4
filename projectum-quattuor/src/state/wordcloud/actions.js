export const getRemoteResource = () => {return {type:'THIS_SHOULD_GET_FROM_REMOTE_IN_THE_FUTURE'}};