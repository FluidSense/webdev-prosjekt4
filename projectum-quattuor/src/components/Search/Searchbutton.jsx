import React,{Component}  from 'react'
import Checkbox from './Checkbox'

const APIQuery = 'https://swapi.co/api/';
const labels =["Planets","Starships","People","Species","Films","Vehicles"];


export class Searchbutton extends Component {
    constructor(props) {
        super(props);
        this.state = {
            endpointValue: '',
            searchValue: '',
            isChecked: false,
            active: '',
        };
    }
    
    /*Funcionality to handle form and state of form*/
    /* Changes state of value whenever the form is changed, in realtime. */
    handleChange = (event) => {
        this.setState({searchValue: event.target.value});
    }

    /* Prevents default formsubmit for now, and logs the state that is saved.*/
    handleSubmit = (event) => {
        event.preventDefault();
    }

    /* Handles state of checkboxes and sets state as to prepend necessary filter for request */
    handleCheck = (event) => {
        this.setState({isChecked: event.target.isChecked});
        this.setState({endpointValue: event.target.value});
        if(this.state.endpointValue === event.target.value){
            this.setState({value: ''})
        }
    }

    /* Creates the checkboxes dynamically from the list of labels. */

    createBoxes() {
        const checkboxArray = []
        const labellength = labels.length;
        for(var i = 0; i < labellength; i++){
           checkboxArray.push(<Checkbox
                key={[i]}
                className="madeBoxes"
                endpointValue={this.state.endpointValue}
                handleChange={this.handleChange}
                handleCheck={this.handleCheck}
                label={labels[i]}
                />)
    }
    return checkboxArray
}

    render() {
        return (
        <div className="search_content">
        <div className="search_wrapper">
        <form onSubmit={this.handleSubmit} method="#">
        <label>
                <input type="text" className="search_bar" value={this.state.searchValue} onChange={this.handleChange} />
            </label>
            <div>
                <input type="submit" className="search_button" value="May the Force be with you." onClick={() =>  this.props.searchWithApi(APIQuery + this.state.endpointValue + this.state.searchValue)}/>
            </div>
        </form> 
           
        </div>
    
        <div className="checkboxes">  
            {this.createBoxes(this.labels)}
        </div>

        <div className="sort_filters"> {/*These are options that the user can make in order to sort and filter the results. The idea is to make it so that changing the value will automatically perform a new request for the result set.*/}
            {/*For sorting the returned objects based on user choice*/}
            <a href="#" className="sort_button">Choose sort method</a>
            <ul className="sorting">
                <li className="sort_optn" href="#" value="lexicographical">Alphabetically</li>
                <li className="sort_optn" href="#" value="by_added_date">By added date</li>
                <li className="sort_optn" href="#" value="by_added_date_rev">By added date reversed</li>
            </ul>
        </div>
        </div>
        );
    }
}
