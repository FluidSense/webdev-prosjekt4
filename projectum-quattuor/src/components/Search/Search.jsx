import React, { Component } from 'react';
import Checkbox from './Checkbox';
import Searchfield from './Searchfield';

const searchLabels = ['Planets', 'Starships', 'People', 'Species', 'Films', 'Vehicles'];
const APIQuery = 'http://it2810-06.idi.ntnu.no/api/api/';
const sortLabels = [['Alphabetical', 'ASC'], ['Alphabetical', 'DESC'], ['ID', 'ASC'], ['ID', 'DESC']];


export default class Searchbutton extends Component {
  constructor(props) {
    super(props);
    this.state = {
      endpointValue: 'planets/',
      selectedItem: 'Planets',
      searchValue: '',
      selectedSort: sortLabels[0],
    };
  }

  /* Funcionality to handle form and state of form */
  /* Changes state of value whenever the form is changed, in realtime. */

  /* Handles state of checkboxes and sets state as to prepend necessary filter for request */
  handleOptionCheck(event, item) {
    this.setState({ endpointValue: event.target.value, selectedItem: item });
  }

  handleFilterCheck(item) {
    this.setState({ selectedSort: item });
  }

  /* Creates the checkboxes dynamically from the list of labels. */

  createOptions() {
    const checkboxArray = [];
    searchLabels.map(item => checkboxArray.push(
      <Checkbox
        key={item}
        className="madeBoxes"
        subKey={item}
        checked={item === this.state.selectedItem}
        handleCheck={e => this.handleOptionCheck(e, item)}
        label={item}
      />,
    ));
    return checkboxArray;
  }

  createSort() {
    const sortArray = [];
    sortLabels.map(item => sortArray.push(
      <Checkbox
        key={item}
        className="madeBoxes"
        subKey={item}
        checked={item === this.state.selectedSort}
        handleCheck={() => this.handleFilterCheck(item)}
        label={`${item[0]} ${item[1]}`}
      />,
    ));
    return sortArray;
  }


  makeNewSearch() {
    this.props.searchWithApi(`${APIQuery}${this.state.endpointValue}?search=${this.state.searchValue}&sortBy=${this.state.selectedSort[0]}&order=${this.state.selectedSort[1]}`); // Request content from search
    this.props.searchForHistoryApi(`${APIQuery}search`); // Request search history
  }

  updateSearchValue(searchValue) {
    this.setState({ searchValue }, () => this.makeNewSearch());
  }

  render() {
    return (
      <div className="search_content">
        <Searchfield
          updateSearchValue={search => this.updateSearchValue(search)}
        />

        <div className="checkboxes">
          {this.createOptions()}
        </div>

        <div className="checkboxes">
          {this.createSort()}
        </div>
      </div>
    );
  }
}
