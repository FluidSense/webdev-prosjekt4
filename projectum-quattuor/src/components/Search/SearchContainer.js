import React from 'react';
import { connect } from 'react-redux';
import { fetchSearchData, fetchSearchHistory } from '../../state/search/actions';
import Search from './Search';


// eslint-disable-next-line no-underscore-dangle
export const _SearchContainer = props => (
  <Search
    searchWithApi={props.searchWithApi}
    searchForHistoryApi={props.searchForHistoryApi}
  />);

/* Callable actions as props */
const mapDispatchToProps = dispatch => ({
  searchWithApi: fetchUrl => dispatch(fetchSearchData(fetchUrl)),
  searchForHistoryApi: fetchUrl => dispatch(fetchSearchHistory(fetchUrl)),
});

/* Connects the container to the store */
const SearchContainer = connect(
  null,
  mapDispatchToProps,
)(_SearchContainer);

export default SearchContainer;
