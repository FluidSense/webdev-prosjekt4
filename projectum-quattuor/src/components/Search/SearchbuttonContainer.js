import React from 'react';
import { connect } from 'react-redux';
import { updateResultSet, updateErrorState, updateLoadingState } from '../../state/example/actions';
//import { resultSetSelector, errorSelector, loadingSelector } from '../../state/example/selectors';
import { Searchbutton } from './Searchbutton';


class _SearchbuttonContainer extends React.Component {

    /*receiveResultUpdate = (val) => {
        this.props.pushToResultSet(val);
    }

    receiveErrorUpdate = (val) => {
        this.props.pushToError(val);
    }

    receiveLoadingUpdate = (val) => {
        this.props.pushLoading(val);
    }*/

    componentDidMount() {
        this.props.pushToResultSet('poop');
    }

    render() {
        return <Searchbutton 
        pushToResultSet={this.props.pushToResultSet} 
        pushToError={this.props.pushToError} 
        pushToLoading={this.props.pushLoading}
        />;
    }

}

/* Callable actions as props */
const mapDispatchToProps = dispatch => {
    return {
        pushToResultSet: payload => dispatch(updateResultSet(payload)),
        pushToError: payload => dispatch(updateErrorState(payload)),
        pushToLoading: payload => dispatch(updateLoadingState(payload)),
    }
}

/* Connects the container to the store */
const SearchbuttonContainer = connect(
    mapDispatchToProps
)(_SearchbuttonContainer);

export default SearchbuttonContainer;