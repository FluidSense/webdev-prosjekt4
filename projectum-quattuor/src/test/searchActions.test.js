import * as actions from '../state/search/actions';
import thunk from 'redux-thunk';
import configureMockStore from 'redux-mock-store';
import fetchMock from 'fetch-mock';
import { APIQuery } from '../components/Search/Search';

describe('Search actions', () => {
    it('should create an action to add a search result', () => {
        const result = {};
        const expectedAction = {
            type: 'FETCHED_RESULTS_RECEIVED',
            payload: result,
        };
        expect(actions.updateSearchContentSet(result)).toEqual(expectedAction);
    });

    it('should create an action to save an error', () => {
        const error = {error:''};
        const expectedAction = {
            type: 'FETCH_CONTENT_ERROR_OCCURRED',
            payload: error,
        };
        expect(actions.updateSearchContentErrorState(error)).toEqual(expectedAction);
    });

    it('should create an action to set loading and url', () => {
        const loading = true; 
        const loadingUrl = 'url';
        const expectedAction = {
            type: 'FETCH_CONTENT_LOADING_IN_PROGRESS',
            payload: { loading, loadingUrl},
        };
        expect(actions.updateSearchContentLoadingState(loading, loadingUrl)).toEqual(expectedAction);
    });

    it('should create an action to set search history', () => {
        const history = [];
        const expectedAction = {
            type: 'FETCHED_HISTORY_RECEIVED',
            payload: history,
        };
        expect(actions.updateSearchHistorySet(history)).toEqual(expectedAction);
    });
});

const middlewares = [thunk];
const mockstore = configureMockStore(middlewares);


describe('Search async action', () => {
    afterEach(() => {
        fetchMock.restore()
    });

    it('creates FETCHED_RESULTS_RECEIVED when fetching search has been successfully done', () => {
        fetchMock.getOnce(APIQuery, {
            body: { todos:'help' },
            headers: {'content-type': 'application/json'}
        });

        const expectedActions  = [
            { type: 'FETCH_CONTENT_LOADING_IN_PROGRESS', payload: { loading:true, loadingUrl:APIQuery }},
            { type: 'FETCHED_RESULTS_RECEIVED', payload: param}
        ];
        const store = mockStore({});

        return store.dispatch(actions.fetchSearchData(APIQuery)).then(() => {
            expect(store.getActions()).toEqual(expectedActions);
        })

    })

});