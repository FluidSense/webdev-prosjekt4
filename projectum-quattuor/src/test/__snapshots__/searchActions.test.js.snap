// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search actions should dispatch the correct action 1`] = `
Array [
  Object {
    "payload": Object {
      "loading": true,
      "loadingUrl": "url",
    },
    "type": "FETCH_CONTENT_LOADING_IN_PROGRESS",
  },
]
`;
